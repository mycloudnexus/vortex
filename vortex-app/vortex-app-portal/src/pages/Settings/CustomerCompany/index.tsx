import type { ReactElement } from 'react'
import type { TableProps } from 'antd'

import { Fragment, useState } from 'react'

import { ReactComponent as CCIcon } from '@/assets/icon/customer-company.svg'
import { ReactComponent as CCEmpty } from '@/assets/icon/customer-company-empty.svg'
import { ReactComponent as CCStatus } from '@/assets/icon/status.svg'
import { ReactComponent as CCClose } from '@/assets/icon/close-circle.svg'
import { ReactComponent as CCWarning } from '@/assets/icon/warning-circle.svg'
import { useAppStore } from '@/stores/app.store'
import Text from '@/components/Text'

import { Button, Flex, Typography, Space, notification } from 'antd'
import { useNavigate } from 'react-router-dom'
import { Company, useCompanyStore } from '@/stores/company.store'
import { useForm } from 'antd/es/form/Form'
import CustomerCompanyModal from '../components/CustomerModal'
import { StyledButton, StyledModal, StyledTable, StyledWrapper } from '../components/styled'
import Tooltip from '../components/Tooltip'

const createColumns = (
  handleOpenModify: (record: Company) => void,
  handleOpenActivate: (key: string) => void,
  handleOpenDeactivate: (key: string) => void
): TableProps<Company>['columns'] => [
  {
    title: 'Name',
    dataIndex: 'title',
    key: 'title'
  },
  {
    title: 'ID',
    dataIndex: 'id',
    key: 'id'
  },
  {
    title: 'Short name & URL',
    dataIndex: 'shortName',
    key: 'shortName',
    render: (_, { shortName }) => (
      <Flex gap={5} align='center'>
        {shortName}
        <Tooltip shortName={shortName} color='#FFF' placement='topLeft' />
      </Flex>
    )
  },
  {
    title: 'Status',
    key: 'status',
    dataIndex: 'status',
    render: (_, { status }) => (
      <>
        {status === 'active' ? (
          <Flex gap={2}>
            <CCStatus style={{ fill: '#00B284' }} />
            <Typography.Text>Active</Typography.Text>
          </Flex>
        ) : (
          <Flex gap={2}>
            <CCStatus style={{ fill: '#668B97' }} />
            <Typography.Text>Inactive</Typography.Text>
          </Flex>
        )}
      </>
    )
  },
  {
    title: 'Action',
    key: 'action',
    render: (_, record) => {
      const handleModify = (e: React.MouseEvent<HTMLElement, MouseEvent>, record: Company): void => {
        e.stopPropagation()
        handleOpenModify(record)
      }
      const handleDeactivate = (e: React.MouseEvent<HTMLElement, MouseEvent>, key: string): void => {
        e.stopPropagation()
        handleOpenDeactivate(key)
      }
      const handleActivate = (e: React.MouseEvent<HTMLElement, MouseEvent>, key: string): void => {
        e.stopPropagation()
        handleOpenActivate(key)
      }
      return (
        <Space size='small'>
          {record.status === 'active' ? (
            <Fragment>
              <Button
                onClick={(e) => handleModify(e, record)}
                variant='link'
                type='link'
                color='danger'
                style={{ padding: '4px 0' }}
              >
                Modify
              </Button>
              <Button
                onClick={(e) => handleDeactivate(e, record.key)}
                variant='link'
                type='link'
                color='danger'
                style={{ padding: '4px 0' }}
              >
                Deactivate
              </Button>
            </Fragment>
          ) : (
            <Fragment>
              <Button
                onClick={(e) => handleActivate(e, record.key)}
                variant='link'
                type='link'
                color='danger'
                style={{ padding: '4px 0' }}
              >
                Activate
              </Button>
            </Fragment>
          )}
        </Space>
      )
    }
  }
]

const CustomerCompany = (): ReactElement => {
  const { mainColor } = useAppStore()
  const { companies, addCompany, updateCompanyRecord, updateCompanyStatus } = useCompanyStore()
  const [form] = useForm()
  const [api, contextHolder] = notification.useNotification({ top: 80 })
  const [isUpdateModalOpen, setIsUpdateModalOpen] = useState<boolean>(false)
  const navigate = useNavigate()
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false)
  const [isWarning, setIsWarning] = useState<boolean>(false)
  const [isConfigLogin, setIsConfigLogin] = useState<boolean>(false)
  const [isDeactivate, setIsDeactivate] = useState<boolean>(false)
  const [updateValue, setUpdateValue] = useState<Company>({
    key: '',
    id: '',
    title: '',
    shortName: '',
    status: 'active'
  })
  const [key, setKey] = useState<string>('')

  const handleSuccessDeactivate = (): void => {
    api.success({
      message: 'Customer company name deactivated',
      placement: 'top',
      showProgress: true,
      pauseOnHover: true,
      closeIcon: false,
      duration: 2
    })
  }

  const handleCloseDeactivate = (): void => setIsDeactivate(false)
  const handleOpenDeactivate = (): void => setIsDeactivate(true)
  const handleDeactivateSubmit = (): void => {
    try {
      updateCompanyStatus(key, 'inactive')
      handleCloseDeactivate()
      handleSuccessDeactivate()
    } catch (error) {
      console.log(error)
    }
  }

  const handleActivate = (key: string): void => {
    setKey(key)
  }
  const handleDeactivate = (key: string): void => {
    setKey(key)
    handleOpenDeactivate()
  }

  const openUpdateModal = (record: Company): void => {
    form.setFieldsValue(record)
    setUpdateValue(record)
    setIsUpdateModalOpen(true)
  }
  const closeUpdateModal = (): void => {
    form.resetFields()
    setIsUpdateModalOpen(false)
  }
  const handleUpdate = async (): Promise<void> => {
    try {
      const values = await form.validateFields()
      updateCompanyRecord({
        ...updateValue,
        title: values.title
      })
      closeUpdateModal()
    } catch (error) {
      console.log('Form validation failed:', error)
    }
  }

  const handleSubmitConfigLogin = (): void => {
    setIsConfigLogin(false)
    navigate(`/settings/customer-company/${updateValue.title}`, { state: { record: updateValue } })
  }
  const handleOpenModal = (record: Company): void => {
    setUpdateValue(record)
    setIsConfigLogin(true)
  }
  const handleCloseConfigModal = (): void => setIsConfigLogin(false)

  const handleOpenWarning = (): void => setIsWarning(true)
  const handleCloseWarning = (): void => setIsWarning(false)

  const showModal = () => {
    setIsModalOpen(true)
  }

  const handleOk = async (): Promise<void> => {
    try {
      const values = await form.validateFields()
      const valuesToAdd: Company = {
        id: 'IDEO2333',
        key: (companies.length + 1).toString(),
        title: values.title,
        shortName: values.shortName,
        status: 'active'
      }
      addCompany(valuesToAdd)
      handleCancel()
      handleOpenModal(valuesToAdd)
    } catch (error) {
      console.log('Form validation failed:', error)
    }
  }

  const handleCancel = () => {
    form.resetFields()
    setIsModalOpen(false)
  }

  const handleClick = (): void => {
    if (companies.length > 200) {
      return handleOpenWarning()
    }
    showModal()
  }
  const handleOnRowClick = (record: Company): void => {
    navigate(`/settings/customer-company/${record.title}`, { state: { record: record } })
  }
  return (
    <Flex style={{ width: '70%', margin: '0 auto' }} vertical justify='start' align='center' gap={20}>
      <Flex justify='space-between' style={{ width: '100%' }} align='center'>
        <Flex gap={15} align='center'>
          <StyledWrapper $backgroundColor={mainColor}>
            <CCIcon />
          </StyledWrapper>
          <Typography.Title level={2} style={{ margin: 0 }}>
            Customer Company
          </Typography.Title>
        </Flex>
        <StyledButton variant='solid' $backgroundColor={mainColor} onClick={handleClick} data-testid='add-button'>
          Add customer company
        </StyledButton>
      </Flex>

      <StyledTable
        columns={createColumns(
          (record) => openUpdateModal(record),
          (key) => handleActivate(key),
          (key) => handleDeactivate(key)
        )}
        dataSource={companies}
        pagination={false}
        onRow={(record) => {
          return {
            onClick: () => handleOnRowClick(record)
          }
        }}
        style={{ width: '100%' }}
        locale={{
          emptyText: (
            <Flex align='center' justify='center' vertical>
              <CCEmpty />
              <Text.NormalLarge color='#000'>No customer company</Text.NormalLarge>
            </Flex>
          )
        }}
      />

      <CustomerCompanyModal
        title='Add customer company'
        name='add_customer_company'
        companies={companies}
        form={form}
        handleCancel={handleCancel}
        handleOk={handleOk}
        initialValues={{ name: '', shortName: '' }}
        isModalOpen={isModalOpen}
      />

      <CustomerCompanyModal
        title='Modify Customer company'
        name='modify_customer_company'
        companies={companies}
        form={form}
        handleCancel={closeUpdateModal}
        handleOk={handleUpdate}
        initialValues={{ name: '', shortName: '' }}
        isModalOpen={isUpdateModalOpen}
        type='update'
      />

      <StyledModal
        centered
        title={
          <Flex align='center' gap={5}>
            <CCClose />
            <Text.NormalMedium style={{ fontSize: '16px', fontWeight: 'bold' }}>
              Cannot create more customer company
            </Text.NormalMedium>
          </Flex>
        }
        open={isWarning}
        okText='Got it'
        onOk={handleCloseWarning}
        closable={false}
        cancelButtonProps={{ style: { display: 'none' } }}
        $containerWidth='25rem'
      >
        <Flex gap={10} vertical style={{ marginLeft: '30px' }}>
          <Text.NormalMedium>Maximum 200 customer companies reached.</Text.NormalMedium>
          <Text.NormalMedium>Any problem, please contact with Console Connect support team</Text.NormalMedium>
        </Flex>
      </StyledModal>

      <StyledModal
        centered
        title={
          <Flex align='center' gap={5}>
            <Text.NormalMedium style={{ fontSize: '16px', fontWeight: 'bold' }}>
              Continue to configure for user login?
            </Text.NormalMedium>
          </Flex>
        }
        open={isConfigLogin}
        okText='Yes, continue'
        onOk={handleSubmitConfigLogin}
        onCancel={handleCloseConfigModal}
        closable={false}
        cancelText='Not now'
        $containerWidth='25rem'
      >
        <Flex gap={10} vertical>
          <Text.NormalMedium>
            Users from this customer company cannot login Vortex yet. <br /> You can select user login method for the
            customer company and invite user or configure SSO accordingly
          </Text.NormalMedium>
        </Flex>
      </StyledModal>

      <StyledModal
        centered
        title={
          <Flex align='center' gap={5}>
            <CCWarning />
            <Text.NormalMedium style={{ fontSize: '16px', fontWeight: 'bold' }}>
              Are you sure to deactivate this customer company?
            </Text.NormalMedium>
          </Flex>
        }
        open={isDeactivate}
        okText='Yes, continue'
        onOk={handleDeactivateSubmit}
        onCancel={handleCloseDeactivate}
        closable={false}
        cancelText='Cancel'
        $containerWidth='30rem'
        okButtonProps={{ style: { backgroundColor: '#FF4D4F' } }}
      >
        <Flex gap={10} vertical style={{ marginLeft: '30px' }}>
          <Text.NormalMedium>
            Please ensure there are no any active services, otherwise the customer will still be charged. <br />
            After deactivation, order history will be kept, but you cannot order new service for this customer. <br />
            The customer users cannot login to Vortex either. <br />
            Continue?
          </Text.NormalMedium>
        </Flex>
      </StyledModal>
      {contextHolder}
    </Flex>
  )
}

export default CustomerCompany
