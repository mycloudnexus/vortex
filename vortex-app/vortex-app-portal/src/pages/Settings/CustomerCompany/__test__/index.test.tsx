import { render } from '@testing-library/react'
import CustomerCompany from '..'
import { MemoryRouter } from 'react-router-dom'

jest.mock('@/assets/icon/customer-company.svg', () => ({
  ReactComponent: () => <svg />
}))
jest.mock('@/assets/icon/customer-company-empty.svg', () => ({
  ReactComponent: () => <svg />
}))
jest.mock('@/assets/icon/status.svg', () => ({
  ReactComponent: () => <svg />
}))
jest.mock('@/assets/icon/close-circle.svg', () => ({
  ReactComponent: () => <svg />
}))
jest.mock('@/assets/icon/warning-circle.svg', () => ({
  ReactComponent: () => <svg />
}))
jest.mock('@/assets/icon/info.svg', () => ({
  ReactComponent: () => <svg />
}))
jest.mock('antd/es/form/Form', () => ({
  useForm: jest.fn()
}))
jest.mock('@/stores/app.store', () => ({
  useAppStore: jest.fn().mockReturnValue({ mainColor: '#000' })
}))
jest.mock('@/stores/company.store', () => ({
  useCompanyStore: jest.fn().mockReturnValue({
    companies: [
      {
        key: '1',
        id: '1',
        title: 'Test Company',
        shortName: 'TC',
        status: 'active'
      }
    ],
    addCompany: jest.fn(),
    updateCompanyRecord: jest.fn(),
    updateCompanyStatus: jest.fn()
  })
}))

jest.mock('antd/es/form/Form', () => ({
  useForm: jest.fn().mockReturnValue([{}, jest.fn()])
}))

global.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  addListener: jest.fn(),
  removeListener: jest.fn()
}))

describe('CutomerCompany Page', () => {
  it('should render the page', () => {
    const { baseElement } = render(
      <MemoryRouter>
        <CustomerCompany />
      </MemoryRouter>
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('should render the company name', () => {
    const { getByText } = render(
      <MemoryRouter>
        <CustomerCompany />
      </MemoryRouter>
    )

    expect(getByText('Test Company')).toBeInTheDocument()
  })

  it('should render the company status', () => {
    const { getByText } = render(
      <MemoryRouter>
        <CustomerCompany />
      </MemoryRouter>
    )

    expect(getByText('Active')).toBeInTheDocument()
  })
})
