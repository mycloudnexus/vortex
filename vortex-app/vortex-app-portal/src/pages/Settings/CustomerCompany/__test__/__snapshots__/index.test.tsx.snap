// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CutomerCompany Page should render the page 1`] = `
<body>
  <div>
    <div
      class="ant-flex css-dev-only-do-not-override-ni1kz0 ant-flex-align-center ant-flex-justify-start ant-flex-vertical"
      style="width: 70%; margin: 0px auto; gap: 20px;"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-ni1kz0 ant-flex-align-center ant-flex-justify-space-between"
        style="width: 100%;"
      >
        <div
          class="ant-flex css-dev-only-do-not-override-ni1kz0 ant-flex-align-center"
          style="gap: 15px;"
        >
          <div
            class="sc-fAUdSK cBpvOS"
          >
            <svg />
          </div>
          <h2
            class="ant-typography css-dev-only-do-not-override-ni1kz0"
            style="margin: 0px;"
          >
            Customer Company
          </h2>
        </div>
        <button
          class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined sc-egkSDF gEXRkt"
          data-testid="add-button"
          type="button"
        >
          <span>
            Add customer company
          </span>
        </button>
      </div>
      <div
        class="ant-table-wrapper sc-dntaoT jPeUFk css-dev-only-do-not-override-ni1kz0"
        style="width: 100%;"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-ni1kz0"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-table css-dev-only-do-not-override-ni1kz0"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup />
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Name
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          ID
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Short name & URL
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Status
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Action
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-row ant-table-row-level-0"
                        data-row-key="1"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          Test Company
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          1
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="ant-flex css-dev-only-do-not-override-ni1kz0 ant-flex-align-center"
                            style="gap: 5px;"
                          >
                            TC
                            <span
                              data-testid="mocked-svg"
                            >
                              <svg />
                            </span>
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="ant-flex css-dev-only-do-not-override-ni1kz0"
                            style="gap: 2px;"
                          >
                            <svg />
                            <span
                              class="ant-typography css-dev-only-do-not-override-ni1kz0"
                            >
                              Active
                            </span>
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="ant-space css-dev-only-do-not-override-ni1kz0 ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
                          >
                            <div
                              class="ant-space-item"
                            >
                              <button
                                class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-link ant-btn-color-dangerous ant-btn-variant-link"
                                style="padding: 4px 0px;"
                                type="button"
                              >
                                <span>
                                  Modify
                                </span>
                              </button>
                            </div>
                            <div
                              class="ant-space-item"
                            >
                              <button
                                class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-link ant-btn-color-dangerous ant-btn-variant-link"
                                style="padding: 4px 0px;"
                                type="button"
                              >
                                <span>
                                  Deactivate
                                </span>
                              </button>
                            </div>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
